python-client/src/gabriel_client/source.py

```diff
@@ -0,0 +1,31 @@
+import asyncio
+import logging
+
+logger = logging.getLogger(__name__)
+
+class _Source:
+    def __init__(self, num_tokens):
+        self._num_tokens = num_tokens
+        self._event = asyncio.Event()
```

Comment:
Would it not make more sense to use an asyncio.Semaphore? Event is a barrier that wakes up all listeners and requires a manual 'clear' to make it block again.
    self._sem = asyncio.Semaphore(num_tokens)

Then in return_token,
  self._sem.release()
 
And in get_token,
  await self._sem.aquire()

No more manual tracking of the current available _num_tokens, it will only wake up a single waiter, if nobody is waiting the event doesn't get lost, etc. If you never want to accidentally exceed num_tokens use a BoundedSemaphore. The only thing that you cannot easily get is a 'get_num_tokens', but except for logging purposes any use of that functions is a race condition waiting to happen, for logging it may be enough to log whether sem.acquire() will block or not by logging the result of sem.locked().

Reply:
Agreed. Changed to use semaphore in cf69838c97c9f467fa86066470250dc994a5cd09